### ISSUE TITLE
einsum permutation bug(possibly?)

### ISSUE THREAD

**Issue by Isalia20** [pytorch/pytorch#116402]
### 🐛 Describe the bug

I'm not sure if categorizing this one as a bug is correct, maybe this is the intended behaviour but if it is please point me to where I can read on why it's done this way.

Code to reproduce:
```
import torch

a = torch.rand((3, 4, 10, 5))
print(torch.einsum("ijkl", a).shape)
```
This outputs (3, 4, 10, 5) which is expected. Then after running:
```
print(torch.einsum("ikjl", a).shape)
```
It outputs (3, 10, 4, 5), which is expected as well, it permutated 1st and 2nd dim, all good. After this I ran:
```
print(torch.einsum("i...l", a).shape)
```
Here what I expected was (3, 4, 10, 5), but got (4, 10, 3, 5) which seems strange to me, then I wrote:
```
print(torch.einsum("l...i", a).shape)
```
Here what I expected was (5, 4, 10, 3) but got (4, 10, 5, 3). 

Can someone explain the logic behind how this works? Because if I write `ijkl` as a formula it doesn't permute the dims but if I write `i...l` it permutes the dims which seems somewhat unintuitive to me

### Versions

PyTorch version: 2.1.2
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.1.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: version 3.26.4
Libc version: N/A

Python version: 3.11.7 | packaged by conda-forge | (main, Dec 15 2023, 08:41:36) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-14.1.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] onnx==1.15.0
[pip3] onnx2torch==1.5.13
[pip3] torch==2.1.2
[pip3] torchvision==0.16.2
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] onnx2torch                1.5.13                   pypi_0    pypi
[conda] torch                     2.1.2                    pypi_0    pypi
[conda] torchvision               0.16.2                   pypi_0    pypi

**Comment by zhihanyang2022** [comment]
I found an answer for this question here: https://pytorch.org/docs/stable/generated/torch.einsum.html.

In particular, look for this quote:

> If the output is not explicitly defined with the arrow (‘->’) notation, the ellipsis will come first in the output (left-most dimensions), before the subscript labels that appear exactly once for the input operands.

This explains why

```{python}
print(torch.einsum("i...l", a).shape)
```

gives `torch.Size([4, 10, 3, 5])` but

```{python}
print(torch.einsum("i...l->i...l", a).shape)
```

gives `torch.Size([3, 4, 10, 5])`.

**Comment by Isalia20** [comment]
Got it, thanks. That explains a lot. Will close this thread
